src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -314,54 +352,54 @@ public boolean indexExists(String index) {
     );
   }
 
-  /**
-   * Maps a record to the write request.
-   *
-   * @param request the write request
-   * @param record  the record
-   */
-  private void addToRequestToRecordMap(DocWriteRequest<?> request, SinkRecord record) {
-    if (requestToRecord != null) {
-      requestToRecord.put(request, record);
-    }
-  }
-
   /**
    * Creates a listener with callback functions to handle completed requests for the BulkProcessor.
    *
    * @return the listener
    */
-  private BulkProcessor.Listener buildListener() {
+  private BulkProcessor.Listener buildListener(OffsetTracker offsetTracker) {
     return new Listener() {
       @Override
       public void beforeBulk(long executionId, BulkRequest request) {
-        if (requestToRecord != null && inFlightRequests != null) {
-          List<SinkRecord> sinkRecords = new ArrayList<>(request.requests().size());
-          for (DocWriteRequest<?> req : request.requests()) {
-            sinkRecords.add(requestToRecord.get(req));
-            requestToRecord.remove(req);
-          }
+        if (inFlightRequests != null) {
+          List<SinkRecordAndOffset> sinkRecords = request.requests().stream()
+                  .map(requestToSinkRecord::get)
+                  .collect(toList());
 
           inFlightRequests.put(executionId, sinkRecords);
         }
       }
 
       @Override
       public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+        List<DocWriteRequest<?>> requests = request.requests();
+
+        int itemNumber = 0;
         for (BulkItemResponse bulkItemResponse : response) {
-          handleResponse(bulkItemResponse, executionId);
+          SinkRecordAndOffset record = itemNumber < requests.size()
+                  ? requestToSinkRecord.get(requests.get(itemNumber))
```

Comment:
Am I missing something or is this always getting the first request only? I guess it doesn't really matter since you always call `markProcessed` on all requests bellow. Maybe no need to pipe through the record to `handleResponse` then?

Reply:
Correct, that's leftover code from before we realized we didn't have to wait for reporter futures (because preCommit already does that). Removed.